stdlib/public/runtime/ProtocolConformance.cpp

```diff
@@ -1374,16 +1387,37 @@ checkGenericRequirement(const GenericRequirementDescriptor &req,
   }
 
   case GenericRequirementKind::SameType: {
+    // Same type requirements don't require a valid subject.
+
+    const Metadata *sameType = nullptr;
+
     // Demangle the second type under the given substitutions.
     auto result = swift_getTypeByMangledName(
         MetadataState::Abstract, req.getMangledTypeName(),
         extraArguments.data(), substGenericParam, substWitnessTable);
-    if (result.getError())
+
+    if (!allowsUnresolvedSubject && result.isError()) {
       return *result.getError();
-    auto otherType = result.getType().getMetadata();
+    }
+
+    if (!result.isError()) {
+      sameType = result.getType().getMetadata();
+    }
+
+    // If we don't have one of either the subject type or the same type and we
+    // have the other, then return this as a success. This assumes the given
+    // extra arguments have provided only the required key arguments in which
+    // case some same type constraints may concretize some generic arguments
+    // making them non-key.
+    //
+    // Note: We don't need to check for allowsUnresolvedSubject here because
+    // these can only be null in the case where we do allow it.
+    if ((!subjectType && sameType) || (subjectType && !sameType)) {
+      return llvm::None;
```

Comment:
I suspect this is going to be wrong with a signature like `<T, U, V where T == U, U == V.A, V: P>`. We would unconditionally accept `U == V.A` because U is not canonical. We really do need to demangle U here via the context descriptor, to realize that it is equal to T, and then effectively check `T == V.A`.

Reply:
Code to recover metadata for non-canonical generic parameters appears in `_buildDemanglingForGenericArgs()`, where we immediately then build a demangle tree from the metadata. This should be split up so that you can re-use the same logic to fill in the "gaps" in the generic arguments array (without building a demangle tree also).